BBQ needs to move stack results from a call to their canonical location

JonWBedard opened the pull request:
#### 53be00f7a1fc6ceb136d6b838a100821c99691f0
<pre>
BBQ needs to move stack results from a call to their canonical location
<a href="https://bugs.webkit.org/show_bug.cgi?id=271175">https://bugs.webkit.org/show_bug.cgi?id=271175</a>
<a href="https://rdar.apple.com/124060867">rdar://124060867</a>

Reviewed by Yusuke Suzuki.

Right now we can end up clobbering a value, `X`, on the stack in BBQ because it gets left in a
`StackArgument` `Location` after a call. This breaks when a later call before `X` has been consumed
would set the same `StackArgument` location as `X`. The fix is to just always move stack results
to their canonical location. This is probably fine because stack results are super rare in practice.

* JSTests/wasm/stress/many-calls-results-on-stack.js: Added.
(repeat):
(check):
(async test):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::returnValuesFromCall):

Originally-landed-as: 272448.770@safari-7618-branch (6d311cd7fefc). <a href="https://rdar.apple.com/128550307">rdar://128550307</a>
Canonical link: <a href="https://commits.webkit.org/279316@main">https://commits.webkit.org/279316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57ef9fc9ac1dfcd7b2a3f0f9e471183a1e769899

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32382 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56324 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3768 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43018 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2435 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55143 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30134 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45791 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24146 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27150 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3113 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1927 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46401 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49002 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3270 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57919 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52558 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28186 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3226 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50415 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29406 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46011 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49716 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11589 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30325 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64863 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29160 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12308 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3959,6 +3959,25 @@ void BBQJIT::returnValuesFromCall(Vector<Value, N>& results, const FunctionSigna
+                Location canonicalLocation = canonicalSlot(result);
+                emitMoveMemory(result.type(), returnLocation, canonicalLocation);
+                returnLocation = canonicalLocation;
             }
```
It looks like this got put in the wrong block. This should be the `else` case of `if (returnLocation.isRegister())` not `if (currentBinding.isScratch())`

JonWBedard replied:
Let me correct this and send it back through.

kmiller68 committed:
BBQ needs to move stack results from a call to their canonical location
https://bugs.webkit.org/show_bug.cgi?id=271175
rdar://124060867

Reviewed by Yusuke Suzuki.

Right now we can end up clobbering a value, `X`, on the stack in BBQ because it gets left in a
`StackArgument` `Location` after a call. This breaks when a later call before `X` has been consumed
would set the same `StackArgument` location as `X`. The fix is to just always move stack results
to their canonical location. This is probably fine because stack results are super rare in practice.

* JSTests/wasm/stress/many-calls-results-on-stack.js: Added.
(repeat):
(check):
(async test):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::returnValuesFromCall):

Originally-landed-as: 272448.770@safari-7618-branch (6d311cd7fefc). rdar://128550307
Canonical link: https://commits.webkit.org/279316@main
